False positive in array_init rule with prefix operators

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Bug Report

```bash
$ Pods/SwiftLint/swiftlint lint --path file.swift                                         
Loading configuration from '.swiftlint.yml'
Linting Swift files at path file.swift
Linting 'file.swift' (1/1)
file.swift:1:1: warning: Array Init Violation: Prefer using Array(seq) than seq.map { $0 } to convert a sequence into an Array. (array_init)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.23.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? CocoaPods
* Paste your configuration file:

```yml
opt_in_rules:
  - array_init
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

No

* Which Xcode version are you using (check `xcode-select -p`)?

Xcode 9

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
array.map { !$0 }
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,10 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#2159](https://github.com/realm/SwiftLint/issues/2159)
 
+* Fix false positives on `array_init` rule when using prefix operators.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#1877](https://github.com/realm/SwiftLint/issues/1877)
+
 ## 0.30.0: A New Washer and Dryer Set
 
 #### Breaking

--- a/Rules.md
+++ b/Rules.md
@@ -276,6 +276,16 @@ foo.something { RouteMapper.map($0) }
 
 ```
 
+```swift
+foo.map { !$0 }
+
+```
+
+```swift
+foo.map { /* a comment */ !$0 }
+
+```
+
 </details>
 <details>
 <summary>Triggering Examples</summary>
@@ -325,6 +335,11 @@ foo.something { RouteMapper.map($0) }
 
 ```swift
 ↓foo.map { $0 /* a comment */ }
+
+```
+
+```swift
+↓foo.map { /* a comment */ $0 }
 
 ```
 

--- a/Source/SwiftLintFramework/Rules/Lint/ArrayInitRule.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/ArrayInitRule.swift
@@ -19,7 +19,9 @@ public struct ArrayInitRule: ASTRule, ConfigurationProviderRule, OptInRule, Auto
             "foo.map { ((), $0) }\n",
             "foo.map { $0! }\n",
             "foo.map { $0! /* force unwrap */ }\n",
-            "foo.something { RouteMapper.map($0) }\n"
+            "foo.something { RouteMapper.map($0) }\n",
+            "foo.map { !$0 }\n",
+            "foo.map { /* a comment */ !$0 }\n"
         ],
         triggeringExamples: [
             "↓foo.map({ $0 })\n",
@@ -37,7 +39,8 @@ public struct ArrayInitRule: ASTRule, ConfigurationProviderRule, OptInRule, Auto
             "↓foo.map { elem -> String in\n" +
             "   elem\n" +
             "}\n",
-            "↓foo.map { $0 /* a comment */ }\n"
+            "↓foo.map { $0 /* a comment */ }\n",
+            "↓foo.map { /* a comment */ $0 }\n"
         ]
     )
 
@@ -68,7 +71,8 @@ public struct ArrayInitRule: ASTRule, ConfigurationProviderRule, OptInRule, Auto
             isParameterStyleViolation(file: file, dictionary: dictionary, tokens: tokens),
             let lastToken = tokens.last,
             case let bodyEndPosition = bodyOffset + bodyLength,
-            !containsTrailingContent(lastToken: lastToken, bodyEndPosition: bodyEndPosition, file: file) else {
+            !containsTrailingContent(lastToken: lastToken, bodyEndPosition: bodyEndPosition, file: file),
+            !containsLeadingContent(tokens: tokens, bodyStartPosition: bodyOffset, file: file) else {
                 return []
         }
 
@@ -98,10 +102,38 @@ public struct ArrayInitRule: ASTRule, ConfigurationProviderRule, OptInRule, Auto
                                          file: File) -> Bool {
         let lastTokenEnd = lastToken.offset + lastToken.length
         let remainingLength = bodyEndPosition - lastTokenEnd
-        let nsstring = file.contents.bridge()
         let remainingRange = NSRange(location: lastTokenEnd, length: remainingLength)
-        let remainingTokens = file.syntaxMap.tokens(inByteRange: remainingRange)
-        let ranges = NSMutableIndexSet(indexesIn: remainingRange)
+        return containsContent(inByteRange: remainingRange, file: file)
+    }
+
+    private func containsLeadingContent(tokens: [SyntaxToken],
+                                        bodyStartPosition: Int,
+                                        file: File) -> Bool {
+        let inTokenPosition = tokens.firstIndex(where: { token in
+            SyntaxKind(rawValue: token.type) == .keyword && file.contents(for: token) == "in"
+        })
+
+        let firstToken: SyntaxToken
+        let start: Int
+        if let position = inTokenPosition {
+            let index = tokens.index(after: position)
+            firstToken = tokens[index]
+            let inToken = tokens[position]
+            start = inToken.offset + inToken.length
+        } else {
+            firstToken = tokens[0]
+            start = bodyStartPosition
+        }
+
+        let length = firstToken.offset - start
+        let remainingRange = NSRange(location: start, length: length)
+        return containsContent(inByteRange: remainingRange, file: file)
+    }
+
+    private func containsContent(inByteRange byteRange: NSRange, file: File) -> Bool {
+        let nsstring = file.contents.bridge()
+        let remainingTokens = file.syntaxMap.tokens(inByteRange: byteRange)
+        let ranges = NSMutableIndexSet(indexesIn: byteRange)
 
         for token in remainingTokens {
             ranges.remove(in: NSRange(location: token.offset, length: token.length))

